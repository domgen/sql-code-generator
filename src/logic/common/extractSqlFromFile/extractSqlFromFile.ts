import { extractSqlFromSqlFile } from './extractSqlFromSqlFile';
import { extractSqlFromTsFile } from './extractSqlFromTsFile';

const SUPPORTED_FILE_EXTENSIONS = ['.sql', '.ts'];
export const extractSqlFromFile = async ({
  filePath,
}: {
  filePath: string;
}) => {
  // 1. grab extension from file; check that we support it
  const extension = `.${filePath.split('.').slice(-1)[0]}`;
  if (!SUPPORTED_FILE_EXTENSIONS.includes(extension)) {
    throw new Error(`file extension '.${extension}' is not supported`);
  }

  // 2. grab sql from file, based on file type
  if (extension === '.sql') return extractSqlFromSqlFile({ filePath });
  if (extension === '.ts') return extractSqlFromTsFile({ filePath });
  throw new Error('unexpected'); // should never reach here, since we checked earlier that the extension was supported
};
