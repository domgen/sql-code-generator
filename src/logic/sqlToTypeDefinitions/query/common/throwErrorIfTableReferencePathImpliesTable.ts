/**
 * instead of "select id from user" we're asking the user to "select user.id from user" or "select u.id from user u"
 *
 * rationale:
 * - this makes it so that we don't have to try and figure out the implicit reference
 * - this makes it so that when a user does join to another table, they don't have to add the explicit references after the fact (i.e., makes it easier to extend their queries)
 */
export const throwErrorIfTableReferencePathImpliesTable = ({ referencePath }: { referencePath: string }) => {
  if (referencePath.split('.').length !== 2) {
    throw new Error(
      `table reference of "${referencePath}" does not have the table alias explicitly defined, violating best practice`,
    );
  }
};
