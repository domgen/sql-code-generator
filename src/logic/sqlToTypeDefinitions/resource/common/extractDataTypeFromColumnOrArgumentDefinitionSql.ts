import { DataType } from '../../../../model';

// https://dev.mysql.com/doc/refman/8.0/en/string-types.html
const mysqlStringTypes = [
  'CHAR',
  'VARCHAR',
  'TINYTEXT',
  'TEXT',
  'MEDIUMTEXT',
  'LONGTEXT',
  'ENUM', // note: we treat enums as strings; TODO: https://github.com/uladkasach/sql-code-generator/issues/1
];

// https://dev.mysql.com/doc/refman/8.0/en/numeric-types.html
const mysqlNumberTypes = [
  'INTEGER',
  'INT',
  'SMALLINT',
  'TINYINT',
  'MEDIUMINT',
  'BIGINT',
  'DECIMAL',
  'NUMERIC',
  'FLOAT',
  'DOUBLE',
];

// https://dev.mysql.com/doc/refman/8.0/en/date-and-time-types.html
const mysqlDateTypes = ['DATE', 'TIME', 'DATETIME', 'TIMESTAMP', 'YEAR'];

// https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
const mysqlBinaryTypes = ['BINARY', 'VARBINARY'];

export const extractDataTypeFromColumnOrArgumentDefinitionSql = ({ sql }: { sql: string }) => {
  // 1. cast to upper
  const sqlUpper = sql.toUpperCase();

  // 2. one by one try to see which datatype is being defined in this definition
  if (mysqlStringTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.STRING;
  if (mysqlNumberTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.NUMBER;
  if (mysqlDateTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.DATE;
  if (mysqlBinaryTypes.some((mysqlType) => sqlUpper.includes(` ${mysqlType}`))) return DataType.BUFFER;
  throw new Error(`could not extract data type from '${sql}'`);
};
