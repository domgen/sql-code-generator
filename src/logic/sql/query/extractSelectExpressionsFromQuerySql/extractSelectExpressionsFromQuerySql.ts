import { extractTypeDefinitionFromSelectExpressionSql } from './extractTypeDefinitionFromSelectExpressionSql';

/*
  selection array starts on seeing a "SELECT"

  TODO: handle subqueries: https://github.com/uladkasach/sql-code-generator/issues/2
*/
export const extractSelectExpressionsFromQuerySql = ({ sql }: { sql: string }) => {
  // 1. grab the content of everything between SELECT and FROM
  const everythingAfterSelect = (() => {
    const partsSplitOnSelect = sql.split(/(?:SELECT|select)/g);
    if (partsSplitOnSelect.length === 1) throw new Error('no "select" keyword found; unexpected'); // fail fast
    if (partsSplitOnSelect.length > 2) throw new Error('more than one "select" keyword found; not yet supported'); // TODO: https://github.com/uladkasach/sql-code-generator/issues/2
    return partsSplitOnSelect[1];
  })();
  const everythingBetweenSelectAndFrom = (() => {
    const partsSplitOnFromAfterSelect = everythingAfterSelect.split(/(?:FROM|from)/g);
    if (partsSplitOnFromAfterSelect.length === 1) throw new Error('no "from" keyword found; unexpected'); // fail fast
    if (partsSplitOnFromAfterSelect.length > 2) {
      throw new Error('more than one "from" keyword found; not yet supported'); // TODO: https://github.com/uladkasach/sql-code-generator/issues/2
    }
    return partsSplitOnFromAfterSelect[0];
  })();

  // 2. split everything between select and from on commas - they are our select expressions
  const selectExpressionsSql = everythingBetweenSelectAndFrom.split(',').map((thisSql) => thisSql.trim());

  // 3. case each select expression sql into a type def of QuerySelectExpression
  return selectExpressionsSql.map((thisSql) => extractTypeDefinitionFromSelectExpressionSql({ sql: thisSql }));
};
