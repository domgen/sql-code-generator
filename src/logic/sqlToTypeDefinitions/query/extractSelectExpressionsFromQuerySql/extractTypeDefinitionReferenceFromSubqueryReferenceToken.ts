import { TypeDefinitionReference } from '../../../../domain';
import { SqlSubqueryReference } from '../../../../domain/objects/SqlSubqueryReference';
import { getTokenForSqlSubqueryReference } from '../common/flattenSqlByReferencingAndTokenizingSubqueries/getTokenForSubqueryReference';
import { extractSelectExpressionsFromQuerySql } from './extractSelectExpressionsFromQuerySql';

export const extractTypeDefinitionReferenceFromSubqueryReferenceToken = ({
  subqueryReferenceToken,
  subqueries,
}: {
  subqueryReferenceToken: string;
  subqueries: SqlSubqueryReference[];
}): TypeDefinitionReference => {
  // 1. find the subquery reference object
  const subquery = subqueries.find(
    (subquery) =>
      getTokenForSqlSubqueryReference({ reference: subquery }) ===
      subqueryReferenceToken,
  );
  if (!subquery) {
    throw new Error(
      'subquery reference token found in select expression - but subquery reference definition not present. unexpected',
    ); // fail fast
  }

  // 2. get select expressions from subquery sql
  const cleanedSubquerySql = subquery.sql.replace(/^\(/, '').replace(/\)$/, ''); // if it has leading or closing paren, strip em
  const selectExpressions = extractSelectExpressionsFromQuerySql({
    sql: cleanedSubquerySql,
    inASubquery: true,
  }); // NOTE: this makes this recursive

  // 3. check that the subquery returns exactly one select expression, otherwise this is a syntax error and not valid sql
  if (selectExpressions.length < 1) {
    throw new Error(
      'subquery in select expression must return atleast one value.',
    );
  }
  if (selectExpressions.length > 1) {
    throw new Error(
      'subquery in select expression may not return more than one value. this is invalid sql',
    );
  }

  // 4. the reference type of this expression is the single reference type of that subquery
  return selectExpressions[0].typeReference;
};
