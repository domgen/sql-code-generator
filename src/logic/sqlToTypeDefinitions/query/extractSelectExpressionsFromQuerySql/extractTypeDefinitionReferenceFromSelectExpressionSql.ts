import { SqlSubqueryReference } from '../../../../model/valueObjects/SqlSubqueryReference';
import { TypeDefinitionReference } from '../../../../model/valueObjects/TypeDefinitionReference';
import { throwErrorIfTableReferencePathImpliesTable } from '../common/throwErrorIfTableReferencePathImpliesTable';
import { extractTypeDefinitionReferenceFromSubqueryReferenceToken } from './extractTypeDefinitionReferenceFromSubqueryReferenceToken';

export const extractTypeDefinitionReferenceFromSelectExpression = ({
  sql,
  subqueries,
}: {
  sql: string;
  subqueries: SqlSubqueryReference[];
}) => {
  // 0. try to extract the reference from a subquery, if query references subquery
  const [
    _, // tslint:disable-line no-unused
    subqueryReferenceToken,
  ] = new RegExp(/^(__SSQ:[\w-]+__)(?:\s+(?:[\w\s]*))?$/).exec(sql) ?? [];
  if (subqueryReferenceToken) {
    return extractTypeDefinitionReferenceFromSubqueryReferenceToken({ subqueryReferenceToken, subqueries });
  }

  // 1. try to extract a tableReferencePath, if query references table
  const [
    __, // tslint:disable-line no-unused
    tableReferencePath,
  ] = new RegExp(/^(\w+\.?\w*)(?:\s+(?:[\w\s]*))?$/).exec(sql) ?? [];
  if (tableReferencePath) {
    throwErrorIfTableReferencePathImpliesTable({ referencePath: tableReferencePath }); // check that table reference is explicitly defined, since its best practice
    return new TypeDefinitionReference({
      tableReferencePath,
      functionReferencePath: null,
    });
  }

  // 2. try to extract a functionReferencePath, if query references function
  const [
    ___, // tslint:disable-line no-unused
    referencedFunctionName,
  ] = new RegExp(/(\w+)(?:\([\w\s\.:,\"\|]+\))/).exec(sql) ?? [];
  if (referencedFunctionName) {
    return new TypeDefinitionReference({
      tableReferencePath: null,
      functionReferencePath: `${referencedFunctionName.toLowerCase()}.output`,
    });
  }

  // 3. if couldn't determine type reference with either of the above, throw an error
  throw new Error(`could not extract type definition reference from '${sql}'`);
};
