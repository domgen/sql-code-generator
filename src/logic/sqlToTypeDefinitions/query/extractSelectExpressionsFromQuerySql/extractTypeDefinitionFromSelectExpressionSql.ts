import { TypeDefinitionOfQuerySelectExpression } from '../../../../domain';
import { SqlSubqueryReference } from '../../../../domain/objects/SqlSubqueryReference';
import { extractTypeDefinitionReferenceFromSelectExpression } from './extractTypeDefinitionReferenceFromSelectExpressionSql';

export const extractTypeDefinitionFromSelectExpressionSql = ({
  sql,
  subqueries,
  inASubquery,
}: {
  sql: string;
  subqueries: SqlSubqueryReference[];
  inASubquery: boolean; // used to indicate whether we are currently running on a subquery or not; being in subquery gives us some implicit info
}) => {
  // grab the type reference
  const typeReference = extractTypeDefinitionReferenceFromSelectExpression({
    sql,
    subqueries,
  });

  // grab the alias, if any
  const [__, specifiedAlias] =
    new RegExp(/(?:[\w\.,:\(\)\s\|\']+)(?:\s+)(?:as|AS)(?:\s+)(\w+)/g).exec(
      sql,
    ) ?? []; // tslint:disable-line no-unused

  // define the alias, considering whether alias was specified
  if (
    !specifiedAlias &&
    !!typeReference.functionReferencePath &&
    !inASubquery
  ) {
    // throw error if fn reference and alias not defined, as its best practice to explicitly define it - and we're not going to "infer" it for the user
    throw new Error(
      `
select expressions that reference a function must have an alias defined, per best practice.
  - e.g., \`select concat(n.first, n.last)\` => \`select concat(n.first, n.last) as full_name\`

\`${typeReference.functionReferencePath}\` does not meet this criteria.
`.trim(),
    );
  }
  const alias = (() => {
    if (specifiedAlias) return specifiedAlias;
    if (inASubquery) return '__subquery_placeholder_name__'; // we wont use this alias anywhere, since we're in a subquery and subquery just returns one value; therefore, we can give it a fake name as a placeholder
    if (typeReference.tableReferencePath)
      return typeReference.tableReferencePath.split('.').slice(-1)[0]!;
    throw new Error('could not define alias for sql expression; unexpected');
  })();

  // return the full definition
  return new TypeDefinitionOfQuerySelectExpression({
    alias,
    typeReference,
  });
};
