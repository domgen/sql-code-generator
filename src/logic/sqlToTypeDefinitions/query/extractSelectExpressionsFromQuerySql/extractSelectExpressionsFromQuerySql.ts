import strip from 'sql-strip-comments';

import { castCommasInParensToPipesForTokenSafety } from '../../resource/common/castCommasInParensToPipesForTokenSafety';
import { flattenSqlByReferencingAndTokenizingSubqueries } from '../common/flattenSqlByReferencingAndTokenizingSubqueries';
import { extractTypeDefinitionFromSelectExpressionSql } from './extractTypeDefinitionFromSelectExpressionSql';

/*
  selection array starts on seeing a "SELECT"
*/
export const extractSelectExpressionsFromQuerySql = ({
  sql,
  inASubquery = false, // false by default
}: {
  sql: string;
  inASubquery?: boolean; // used to indicate whether we are currently running on a subquery or not; being in subquery gives us some implicit info
}) => {
  // 0. strip sql comments
  const strippedSql = strip(sql);

  // 1. tokenize each subquery
  const { flattenedSql, references: subqueries } =
    flattenSqlByReferencingAndTokenizingSubqueries({ sql: strippedSql });

  // 2. grab the content of everything between SELECT and FROM
  const everythingAfterSelect = (() => {
    const partsSplitOnSelect = flattenedSql.split(/(?:SELECT|select)/g);
    if (partsSplitOnSelect.length === 1)
      throw new Error('no "select" keyword found; unexpected'); // fail fast
    if (partsSplitOnSelect.length > 2)
      throw new Error('more than one "select" keyword found; unexpected'); // should have been flattened out
    return partsSplitOnSelect[1];
  })();
  const everythingBetweenSelectAndFrom = (() => {
    const partsSplitOnFromAfterSelect = everythingAfterSelect.split(
      /[\n\r\s]+from[\n\r\s]+/gi,
    );
    if (partsSplitOnFromAfterSelect.length === 1) return everythingAfterSelect; // if no "FROM" keyword, then expect that the whole query is just the select expressions
    if (partsSplitOnFromAfterSelect.length > 2) {
      throw new Error('more than one "from" keyword found; unexpected'); // should have been flattened out
    }
    return partsSplitOnFromAfterSelect[0];
  })();

  // 3. cast commas inside of parens into pipes, so that we treat them as unique tokens when splitting "property lines" by comma
  const parenTokenizedInnerSql = castCommasInParensToPipesForTokenSafety({
    sql: everythingBetweenSelectAndFrom,
  });

  // 4. split everything between select and from on commas - they are our select expressions
  const selectExpressionsSql = parenTokenizedInnerSql
    .split(',')
    .map((thisSql) => thisSql.trim());

  // 5. case each select expression sql into a type def of QuerySelectExpression
  return selectExpressionsSql.map((thisSql) =>
    extractTypeDefinitionFromSelectExpressionSql({
      sql: thisSql,
      subqueries,
      inASubquery,
    }),
  );
};
