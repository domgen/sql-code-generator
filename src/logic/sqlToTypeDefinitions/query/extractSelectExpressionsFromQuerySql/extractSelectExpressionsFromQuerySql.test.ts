import { extractSqlFromFile } from '../../../common/extractSqlFromFile';
import { extractSelectExpressionsFromQuerySql } from './extractSelectExpressionsFromQuerySql';

describe('extractSelectExpressionsFromQuerySql', () => {
  it('should be able to determine types accurately for this example', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/find_image_by_id.sql`,
    });
    const defs = extractSelectExpressionsFromQuerySql({ sql });
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for this other example', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/select_suggestion.sql`,
    });
    const defs = extractSelectExpressionsFromQuerySql({ sql });
    expect(defs.slice(-1)[0]!.alias).toEqual('updated_at');
    expect(defs.slice(-1)[0]!.typeReference.tableReferencePath).toEqual(
      'v.created_at',
    );
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for query selecting function output', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/upsert_suggestion.sql`,
    });
    const defs = extractSelectExpressionsFromQuerySql({ sql });
    expect(defs[0]!.alias).toEqual('id');
    expect(defs[0]!.typeReference.functionReferencePath).toEqual(
      'upsert_suggestion.output',
    );
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for query selecting function output and normal table columns', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/find_users_by_last_name.sql`,
    });
    const defs = extractSelectExpressionsFromQuerySql({ sql });
    expect(defs[1]!.alias).toEqual('full_name');
    expect(defs[1]!.typeReference.functionReferencePath).toEqual(
      'concat.output',
    );
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for query with subquery', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../__test_assets__/find_with_subselect_in_select_expressions.sql`,
    });
    const defs = extractSelectExpressionsFromQuerySql({ sql });
    expect(defs[3]!.alias).toEqual('ingredient_ids');
    expect(defs[3]!.typeReference.functionReferencePath).toEqual(
      'group_concat.output',
    );
    expect(defs).toMatchSnapshot();
  });
});
