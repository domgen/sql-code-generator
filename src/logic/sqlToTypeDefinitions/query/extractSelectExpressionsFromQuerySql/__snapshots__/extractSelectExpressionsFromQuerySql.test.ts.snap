// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for query selecting function output 1`] = `
Array [
  TypeDefinitionOfQuerySelectExpression {
    "alias": "id",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": "upsert_suggestion.output",
      "tableReferencePath": null,
    },
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for query selecting function output and normal table columns 1`] = `
Array [
  TypeDefinitionOfQuerySelectExpression {
    "alias": "id",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "u.id",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "full_name",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": "concat.output",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "age",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "u.age",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "phone_number",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "p.number",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "description",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "up.description",
    },
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for query with subquery 1`] = `
Array [
  TypeDefinitionOfQuerySelectExpression {
    "alias": "id",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "s.id",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "uuid",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "s.uuid",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "name",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "s.name",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "ingredient_ids",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": "group_concat.output",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "created_at",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "s.created_at",
    },
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for this example 1`] = `
Array [
  TypeDefinitionOfQuerySelectExpression {
    "alias": "id",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "i.id",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "url",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "i.url",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "caption",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "i.caption",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "alt_text",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "i.alt_text",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "credit",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "i.credit",
    },
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for this other example 1`] = `
Array [
  TypeDefinitionOfQuerySelectExpression {
    "alias": "id",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "s.id",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "uuid",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "s.uuid",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "suggestion_source",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "s.suggestion_source",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "external_id",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "s.external_id",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "suggested_idea_id",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "s.suggested_idea_id",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "resultant_curated_idea_id",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "s.resultant_curated_idea_id",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "status",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "v.status",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "result",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "v.result",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "created_at",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "s.created_at",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "effective_at",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "v.effective_at",
    },
  },
  TypeDefinitionOfQuerySelectExpression {
    "alias": "updated_at",
    "typeReference": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "v.created_at",
    },
  },
]
`;
